movement/io/validators.py

```diff
@@ -1,39 +1,348 @@
+        list (default), this check is skipped.
+
+    Raises
+    ------
+    ValueError
+        If the file is not in HDF5 format or if it does not contain the
+        expected datasets.
+    """
+
+    path: Path = field(validator=validators.instance_of(Path))
+    expected_datasets: List[str] = field(factory=list, kw_only=True)
+
+    @path.validator
+    def file_is_h5(self, attribute, value):
+        """Ensure that the file is indeed in HDF5 format."""
+        try:
+            with h5py.File(value, "r") as f:
+                f.close()
+        except Exception as e:
+            raise log_and_raise_error(
+                ValueError,
+                f"File {value} does not seem to be in valid" "HDF5 format.",
+            ) from e
+
+    @path.validator
+    def file_contains_expected_datasets(self, attribute, value):
+        """Ensure that the HDF5 file contains the expected datasets."""
+        if self.expected_datasets:
+            with h5py.File(value, "r") as f:
+                diff = set(self.expected_datasets).difference(set(f.keys()))
+                if len(diff) > 0:
+                    raise log_and_raise_error(
+                        ValueError,
+                        f"Could not find the expected dataset(s) {diff} "
+                        f"in file: {value}. ",
+                    )
+
+
+@define
+class ValidPosesCSV:
+    """Class for validating CSV files that contain pose estimation outputs.
+    in DeepLabCut format.
+
+    Parameters
+    ----------
+    path : pathlib.Path
+        Path to the CSV file.
+    multianimal : bool
+        Whether to ensure that the CSV file contains pose estimation outputs
+        for multiple animals. Default: False.
+
+    Raises
+    ------
+    ValueError
+        If the CSV file does not contain the expected DeepLabCut index column
+        levels among its top rows.
+    """
+
+    path: Path = field(validator=validators.instance_of(Path))
+    multianimal: bool = field(default=False, kw_only=True)
+
+    @path.validator
+    def csv_file_contains_expected_levels(self, attribute, value):
+        """Ensure that the CSV file contains the expected index column levels
+        among its top rows."""
+        expected_levels = ["scorer", "bodyparts", "coords"]
+        if self.multianimal:
+            expected_levels.insert(1, "individuals")
+
+        with open(value, "r") as f:
+            header_rows_start = [f.readline().split(",")[0] for _ in range(4)]
+            level_in_header_row_starts = [
+                level in header_rows_start for level in expected_levels
+            ]
+            if not all(level_in_header_row_starts):
+                raise log_and_raise_error(
+                    ValueError,
+                    f"The header rows of the CSV file {value} do not "
+                    "contain all expected index column levels "
+                    f"{expected_levels}.",
+                )
```

Comment:
Thanks for this. I like your approach and wanted to adopt it, but the tests failed. This forced me to rethink my whole approach for this particular validator.

The fundamental problem was that I had a boolean attribute `multianimal`, to signify whether I was expecting a single- or a multi-animal DLC files (the latter has an extra column level). However, I cannot possibly know if I file will be single- or multi- before reading it, and I don't want to force the user to tell me. I therefore rewrote the validator thus:

```python
@path.validator
def csv_file_contains_expected_levels(self, attribute, value):
      """Ensure that the CSV file contains the expected index column levels
      among its top rows."""
      expected_levels = ["scorer", "bodyparts", "coords"]

      with open(value, "r") as f:
          top4_row_starts = [f.readline().split(",")[0] for _ in range(4)]

          if top4_row_starts[3].isdigit():
              # if 4th row starts with a digit, assume single-animal DLC file
              expected_levels.append(top4_row_starts[3])
          else:
              # otherwise, assume multi-animal DLC file
              expected_levels.insert(1, "individuals")

              if top4_row_starts != expected_levels:
                  raise log_error(
                      ValueError,
                      "CSV header rows do not match the known format for "
                      "DeepLabCut pose estimation output files.",
                  )
```

With this approach, I try to infer the type of the file. If the 4th row starts with a digit (frame number) then the CSV can only be a single-animal file, and has to match the corresponding 3 levels. Otherwise, it should match the levels of a multi-animal file. 
Furthermore, instead of simply checking if a string (e.g. "coords") is among the expected levels, I now check for list equality, which is stricter, and also ensures that the levels come in the right order.

Let me know what you think.

Reply:
Nice! was an oversight from me that the tests were failing with my proposed solution.